### PR TITLE
fix(autoselect): demote proxies reset mid-stream

### PR DESCRIPTION
**Pull request by garmr-ulfr**
This pull request improves the robustness and clarity of failure detection and attribution in the data plane watchdog logic, and significantly enhances the test coverage for mid-stream errors and edge cases. The main changes include introducing explicit mid-stream transport failure attribution, updating the handling of benign terminal conditions, and refactoring tests for better reliability and coverage.

**Data Plane Failure Detection and Attribution**

- Added a new `isDataPlaneFailure` function to distinguish between real transport failures (such as connection resets) and benign conditions (like EOF, context cancellation, or timeouts), ensuring only genuine failures demote the outbound connection.
- Introduced the `fireFailure` method in `dataPlaneWatchdog`, which immediately attributes a failure on explicit mid-stream transport errors, bypassing the proven/lastWasWrite gates and ensuring accurate and timely demotion.

**Improvements to Idle and Error Handling Logic**

- Updated `noteIO` to use the new failure detection logic, ensuring that empty I/O and benign errors do not trigger failure attribution, and that attribution is short-circuited after the connection is already marked as stalled.
- Clarified the documentation and logic for `fireStall` to better explain its role in demoting connections that go idle mid-stream.

**Test Suite Enhancements**

- Added comprehensive new test cases covering mid-stream resets, timeouts, EOF, error attribution after close, and ensuring that failures are attributed at most once per connection. This includes new helper types like `failingConn` and `failingPacketConn` for simulating error scenarios.
- Refactored existing tests to use `require.NoError` and `require.FailNow` for more consistent and reliable assertions, replacing manual error checks and `t.Fatal` calls. [[1]](diffhunk://#diff-99236f6871d4d70c9c587ac714d931de3cbf86939d72bbfee20b7cbdb85f8d57R578-R588) [[2]](diffhunk://#diff-99236f6871d4d70c9c587ac714d931de3cbf86939d72bbfee20b7cbdb85f8d57L604-R611) [[3]](diffhunk://#diff-99236f6871d4d70c9c587ac714d931de3cbf86939d72bbfee20b7cbdb85f8d57L834-R841) [[4]](diffhunk://#diff-99236f6871d4d70c9c587ac714d931de3cbf86939d72bbfee20b7cbdb85f8d57L849-R856) [[5]](diffhunk://#diff-99236f6871d4d70c9c587ac714d931de3cbf86939d72bbfee20b7cbdb85f8d57L992-R1190) [[6]](diffhunk://#diff-99236f6871d4d70c9c587ac714d931de3cbf86939d72bbfee20b7cbdb85f8d57L1009-R1211) [[7]](diffhunk://#diff-99236f6871d4d70c9c587ac714d931de3cbf86939d72bbfee20b7cbdb85f8d57L1030-R1228) [[8]](diffhunk://#diff-99236f6871d4d70c9c587ac714d931de3cbf86939d72bbfee20b7cbdb85f8d57L1063-R1261) [[9]](diffhunk://#diff-99236f6871d4d70c9c587ac714d931de3cbf86939d72bbfee20b7cbdb85f8d57L1118-R1316)

These changes collectively make the data plane watchdog more accurate in detecting real failures, prevent spurious demotions, and ensure that the test suite reliably covers a wide range of edge cases.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of data-plane transport errors so only real mid-stream failures trigger demotion, while benign endings like EOF, cancellation, and timeouts are ignored.
  * Prevented stall/failure callbacks from blocking by running them asynchronously and ensuring they fire only once.

* **Tests**
  * Expanded and refined coverage for data-plane stall behavior, failure-attribution rules, and exhaustion signal delivery.

### CODE REVIEW

File: protocol/group/mutableautoselect_dataplane.go [hunk 1]
```diff
@@ -62,15 +65,35 @@ func (w *dataPlaneWatchdog) init(idle time.Duration, provedReadBytes uint64, onS
...
+func isDataPlaneFailure(err error) bool {
+	if err == nil || errors.Is(err, io.EOF) || errors.Is(err, context.Canceled) {
+		return false
 	}
+	var ne net.Error
+	if errors.As(err, &ne) && ne.Timeout() {
+		return false
+	}
+	return true
+}
```
[the remark this answers is not in the document]
Comment (garmr-ulfr) [reply protocol/group/mutableautoselect_dataplane.go:79]: `context.DeadlineExceeded` is already handled by the `net.Error` branch, so it does not demote the outbound. Its concrete type (`context.deadlineExceededError`) implements `Error()`, `Timeout() bool` (returns `true`), and `Temporary() bool`, which satisfies the `net.Error` interface — so `errors.As(err, &ne) && ne.Timeout()` matches it and returns `false` (benign).

Verified:
```
errors.As(context.DeadlineExceeded, &net.Error) = true
ne.Timeout() = true
treated as benign (not a failure) = true
```

The explicit `errors.Is(err, context.Canceled)` check covers the one cancellation error that is *not* a `net.Error` (`context.Canceled` has no `Timeout()` method). Resolving as a false positive.

**Review by WendelHime — APPROVED** [review]
